File: experiments/lemonade-wars/src/lemonwars/player.py
"""Player state: cash, loans, inventory, equipment, staff, reputation.

The player owns no pygame or UI state — this is pure simulation state so the
whole game can run headless (tests, autoplay) without a display.
"""

from __future__ import annotations

import random

from . import data


class Loan:
    __slots__ = ("principal", "rate")

    def __init__(self, principal: float, rate: float):
        self.principal = float(principal)
        self.rate = float(rate)


class Player:
    def __init__(self, difficulty: data.Difficulty):
        # The $500 starting cash plus the proceeds of the $2,000 business loan
        # (the loan itself is recorded as a liability below).
        self.cash: float = round(difficulty.start_cash + data.START_LOAN, 2)
        self.loans: list[Loan] = [Loan(data.START_LOAN, data.START_LOAN_RATE)]
        self.inventory: dict[str, dict] = {}       # ing -> {"qty": float, "age": int}
        self.products: dict[str, dict] = {}        # key -> {"qty": float, "age": int}
        self.equipment: set[str] = set()
        self.tech: set[str] = set()
        self.tier_idx: int = 0
        self.insurance: set[str] = set()
        self.employees: dict[str, dict] = {}       # role -> {"skill","morale","days"}
        self.awareness: float = 5.0
        self.reputation: float = data.START_REPUTATION
        # weekly finance
        self.week_profit: float = 0.0
        self.sales_tax_bill: float = 0.0
        self.payroll_accum: float = 0.0
        self.cleanliness: float = 60.0
        self.district: str = data.START_DISTRICT
        self.visited: set[str] = {data.START_DISTRICT}

    # -- money ---------------------------------------------------------------

    @property
    def debt(self) -> float:
        return sum(l.principal for l in self.loans)

    def tier(self) -> data.Tier:
        return data.TIERS[self.tier_idx]

    def capacity(self) -> int:
        cap = self.tier().storage
        if "squeezer" in self.equipment:
            cap += 10
        if "juicer" in self.equipment:
            cap += 15
        if "delivery_vehicle" in self.equipment:
            cap += 25
        if "storefront" in self.equipment:
            cap += 100
        return cap

    def storage_used(self) -> float:
        return sum(s["qty"] for s in self.inventory.values()) + \
               sum(s["qty"] for s in self.products.values())

    def storage_room(self) -> float:
        return max(0.0, self.capacity() - self.storage_used())

    # -- inventory -----------------------------------------------------------

    def stock(self, ing: str) -> float:
        return self.inventory.get(ing, {}).get("qty", 0.0)

    def product_stock(self, key: str) -> float:
        return self.products.get(key, {}).get("qty", 0.0)

    def buy_ingredient(self, ing: str, qty: float, price: float, room: float = 1e9) -> bool:
        qty = round(float(qty), 3)
        if qty <= 0:
            return False
        cost = round(qty * price, 2)
        if cost > self.cash + 1e-6:
            return False
        if qty > room + 1e-6:
            return False
        self.cash = round(self.cash - cost, 2)
        stock = self.inventory.setdefault(ing, {"qty": 0.0, "age": 0})
        stock["qty"] = round(stock["qty"] + qty, 3)
        return True

    def consume(self, ing: str, qty: float) -> bool:
        stock = self.inventory.get(ing)
        if not stock or stock["qty"] < qty - 1e-6:
            return False
        stock["qty"] = round(stock["qty"] - qty, 3)
        if stock["qty"] <= 0.01:
            self.inventory.pop(ing, None)
        return True

    def add_products(self, key: str, qty: float) -> None:
        stock = self.products.setdefault(key, {"qty": 0.0, "age": 0})
        stock["qty"] = round(stock["qty"] + qty, 3)

    def ingredient_freshness(self, product: data.Product) -> float:
        """0 (all fresh) .. 1 (all about to turn) weighted over the recipe."""
        ing = data.INGREDIENTS
        used = product.base_recipe
        total = 0.0
        weight_sum = 0.0
        for k, units in used.items():
            if k not in self.inventory or self.inventory[k]["qty"] <= 0:
                total += 1.0
                weight_sum += units
                continue
            age = self.inventory[k]["age"]
            shelf = max(1, ing[k].shelf)
            total += min(1.0, age / shelf) * units
            weight_sum += units
        return total / max(weight_sum, 1e-6)

    def spoilage(self, game=None) -> float:
        """Age everything one day; return units lost (stats)."""
        lost = 0.0
        ing = data.INGREDIENTS
        heat = game is not None and game.world.weather_at(self.district).key == "heat_wave"
        has_fridge = "refrigeration" in self.equipment
        automation = "inventory_automation" in self.tech

        for k, s in list(self.inventory.items()):
            s["age"] += 1
            shelf = ing[k].shelf
            if s["age"] > shelf:
                lost += s["qty"]
                self.inventory.pop(k)
                continue
            ratio = s["age"] / shelf
            loss = 0.0
            if heat and not has_fridge and k in ("ice", "milk", "strawberries", "blueberries", "mango", "organic_fruit", "exotic_fruit"):
                loss += 0.05
            if ratio > 0.85 and not has_fridge:
                loss += 0.02
            if loss > 0:
                drop = round(s["qty"] * loss, 3)
                if automation:
                    drop *= 0.8
                s["qty"] = round(s["qty"] - drop, 3)
                lost += drop
                if s["qty"] <= 0.01:
                    self.inventory.pop(k)

        for k, s in list(self.products.items()):
            s["age"] += 1
            shelf = 2 + (1 if has_fridge else 0)
            if k in ("coffee", "cold_brew"):
                shelf += 1
            if s["age"] > shelf:
                lost += s["qty"]
                self.products.pop(k)
        return lost

    # -- valuation -----------------------------------------------------------

    def inventory_value(self, market) -> float:
        value = 0.0
        for k, s in self.inventory.items():
            value += s["qty"] * market.price(self.district, k)
        for k, s in self.products.items():
            value += s["qty"] * data.PRODUCTS[k].anchor * 0.7
        return round(value, 2)

    def equipment_value(self) -> float:
        return round(sum(data.EQUIPMENT[k].cost * 0.6 for k in self.equipment), 2)

    def tier_value(self) -> float:
        return round(sum(t.cost for t in data.TIERS[: self.tier_idx]), 2)

    def net_worth(self, market) -> float:
        return round(self.cash + self.inventory_value(market) +
                     self.equipment_value() + self.tier_value() - self.debt, 2)

    # -- employees -----------------------------------------------------------

    def hire(self, role: str, rng: random.Random) -> bool:
        if role in self.employees or role not in data.EMPLOYEE_ROLES:
            return False
        self.employees[role] = {
            "skill": round(rng.uniform(30, 60), 1),
            "morale": 100.0,
            "days": 0,
        }
        return True

    def fire(self, role: str) -> bool:
        return self.employees.pop(role, None) is not None

    def daily_wages(self) -> float:
        return round(sum(data.EMPLOYEE_ROLES[r].salary for r in self.employees), 2)

    def staff_bonus(self, kind: str) -> float:
        """kind in service | quality | cleanliness | fuel. Pure — no mutation."""
        vals = {
            "cashier": {"service": 12, "quality": 0, "cleanliness": 0},
            "cook": {"service": 0, "quality": 6, "cleanliness": 0},
            "cleaner": {"service": 0, "quality": 0, "cleanliness": 20},
            "driver": {"service": 0, "quality": 0, "cleanliness": 0},
        }
        total = 0.0
        for role, e in self.employees.items():
            eff = (e["skill"] / 50.0) * (0.7 + 0.3 * e["morale"] / 100.0)
            total += vals[role][kind] * eff
        return total

    def tick_staff(self) -> None:
        """Advance staff days-worked and decay morale — once per day."""
        for e in self.employees.values():
            e["morale"] = max(0.0, e["morale"] - 0.8)
            e["days"] += 1

    def snapshot(self) -> dict:
        return {
            "cash": self.cash,
            "loans": [{"principal": l.principal, "rate": l.rate} for l in self.loans],
            "inventory": {k: dict(v) for k, v in self.inventory.items()},
            "products": {k: dict(v) for k, v in self.products.items()},
            "equipment": sorted(self.equipment),
            "tech": sorted(self.tech),
            "tier_idx": self.tier_idx,
            "insurance": sorted(self.insurance),
            "employees": {k: dict(v) for k, v in self.employees.items()},
            "awareness": self.awareness,
            "reputation": self.reputation,
            "week_profit": self.week_profit,
            "sales_tax_bill": self.sales_tax_bill,
            "payroll_accum": self.payroll_accum,
            "cleanliness": self.cleanliness,
            "district": self.district,
            "visited": sorted(self.visited),
        }

    @classmethod
    def restore(cls, snap: dict, difficulty: data.Difficulty) -> "Player":
        p = cls(difficulty)
        p.cash = float(snap["cash"])
        p.loans = [Loan(l["principal"], l["rate"]) for l in snap["loans"]]
        p.inventory = {k: {"qty": float(v["qty"]), "age": int(v["age"])} for k, v in snap["inventory"].items()}
        p.products = {k: {"qty": float(v["qty"]), "age": int(v["age"])} for k, v in snap["products"].items()}
        p.equipment = set(snap["equipment"])
        p.tech = set(snap["tech"])
        p.tier_idx = int(snap["tier_idx"])
        p.insurance = set(snap["insurance"])
        p.employees = {k: dict(v) for k, v in snap["employees"].items()}
        p.awareness = float(snap["awareness"])
        p.reputation = float(snap["reputation"])
        p.week_profit = float(snap["week_profit"])
        p.sales_tax_bill = float(snap["sales_tax_bill"])
        p.payroll_accum = float(snap["payroll_accum"])
        p.cleanliness = float(snap["cleanliness"])
        p.district = snap["district"]
        p.visited = set(snap["visited"])
        return p

File: experiments/lemonade-wars/js/player.js
// Player state — direct port of player.py: cash, loans, inventory,
// equipment, staff, reputation. Pure simulation state, no UI.
"use strict";

import { EQUIPMENT, INGREDIENTS, PRODUCTS, START_DISTRICT, START_LOAN, START_LOAN_RATE, START_REPUTATION, TIERS } from "./data.js";

const round2 = (v) => Math.round(v * 100) / 100;
const round3 = (v) => Math.round(v * 1000) / 1000;

export class Loan {
  constructor(principal, rate) {
    this.principal = Number(principal);
    this.rate = Number(rate);
  }
}

export class Player {
  constructor(difficulty) {
    // The starting cash plus the proceeds of the $2,000 business loan
    // (the loan itself is recorded as a liability).
    this.cash = round2(difficulty.start_cash + START_LOAN);
    this.loans = [new Loan(START_LOAN, START_LOAN_RATE)];
    this.inventory = {}; // ing -> {qty, age}
    this.products = {}; // key -> {qty, age}
    this.equipment = new Set();
    this.tech = new Set();
    this.tier_idx = 0;
    this.insurance = new Set();
    this.employees = {}; // role -> {skill, morale, days}
    this.awareness = 5.0;
    this.reputation = START_REPUTATION;
    // weekly finance
    this.week_profit = 0.0;
    this.sales_tax_bill = 0.0;
    this.payroll_accum = 0.0;
    this.cleanliness = 60.0;
    this.district = START_DISTRICT;
    this.visited = new Set([START_DISTRICT]);
  }

  // -- money ---------------------------------------------------------------

  get debt() {
    return this.loans.reduce((s, l) => s + l.principal, 0);
  }

  tier() {
    return TIERS[this.tier_idx];
  }

  capacity() {
    let cap = this.tier().storage;
    if (this.equipment.has("squeezer")) cap += 10;
    if (this.equipment.has("juicer")) cap += 15;
    if (this.equipment.has("delivery_vehicle")) cap += 25;
    if (this.equipment.has("storefront")) cap += 100;
    return cap;
  }

  storageUsed() {
    let used = 0;
    for (const s of Object.values(this.inventory)) used += s.qty;
    for (const s of Object.values(this.products)) used += s.qty;
    return used;
  }

  storageRoom() {
    return Math.max(0.0, this.capacity() - this.storageUsed());
  }

  // -- inventory -----------------------------------------------------------

  stock(ing) {
    return this.inventory[ing]?.qty ?? 0.0;
  }

  productStock(key) {
    return this.products[key]?.qty ?? 0.0;
  }

  buyIngredient(ing, qty, price, room = 1e9) {
    qty = round3(Number(qty));
    if (qty <= 0) return false;
    const cost = round2(qty * price);
    if (cost > this.cash + 1e-6) return false;
    if (qty > room + 1e-6) return false;
    this.cash = round2(this.cash - cost);
    const stock = (this.inventory[ing] ??= { qty: 0.0, age: 0 });
    stock.qty = round3(stock.qty + qty);
    return true;
  }

  consume(ing, qty) {
    const stock = this.inventory[ing];
    if (!stock || stock.qty < qty - 1e-6) return false;
    stock.qty = round3(stock.qty - qty);
    if (stock.qty <= 0.01) delete this.inventory[ing];
    return true;
  }

  addProducts(key, qty) {
    const stock = (this.products[key] ??= { qty: 0.0, age: 0 });
    stock.qty = round3(stock.qty + qty);
  }

  ingredientFreshness(product) {
    const used = product.base_recipe;
    let total = 0.0;
    let weightSum = 0.0;
    for (const [k, units] of Object.entries(used)) {
      const inv = this.inventory[k];
      if (!inv || inv.qty <= 0) {
        total += 1.0;
        weightSum += units;
        continue;
      }
      const shelf = Math.max(1, INGREDIENTS[k].shelf);
      total += Math.min(1.0, inv.age / shelf) * units;
      weightSum += units;
    }
    return total / Math.max(weightSum, 1e-6);
  }

  spoilage(game = null) {
    // Age everything one day; return units lost (stats).
    let lost = 0.0;
    const heat = game !== null && game.world.weatherAt(this.district).key === "heat_wave";
    const hasFridge = this.equipment.has("refrigeration");
    const automation = this.tech.has("inventory_automation");
    const chilled = new Set(["ice", "milk", "strawberries", "blueberries", "mango", "organic_fruit", "exotic_fruit"]);

    for (const [k, s] of Object.entries(this.inventory)) {
      s.age += 1;
      const shelf = INGREDIENTS[k].shelf;
      if (s.age > shelf) {
        lost += s.qty;
        delete this.inventory[k];
        continue;
      }
      const ratio = s.age / shelf;
      let loss = 0.0;
      if (heat && !hasFridge && chilled.has(k)) loss += 0.05;
      if (ratio > 0.85 && !hasFridge) loss += 0.02;
      if (loss > 0) {
        let drop = round3(s.qty * loss);
        if (automation) drop *= 0.8;
        s.qty = round3(s.qty - drop);
        lost += drop;
        if (s.qty <= 0.01) delete this.inventory[k];
      }
    }

    for (const [k, s] of Object.entries(this.products)) {
      s.age += 1;
      let shelf = 2 + (hasFridge ? 1 : 0);
      if (k === "coffee" || k === "cold_brew") shelf += 1;
      if (s.age > shelf) {
        lost += s.qty;
        delete this.products[k];
      }
    }
    return lost;
  }

  // -- valuation -----------------------------------------------------------

  inventoryValue(market) {
    let value = 0.0;
    for (const [k, s] of Object.entries(this.inventory)) {
      value += s.qty * market.price(this.district, k);
    }
    for (const [k, s] of Object.entries(this.products)) {
      value += s.qty * PRODUCTS[k].anchor * 0.7;
    }
    return round2(value);
  }

  equipmentValue() {
    let v = 0;
    for (const k of this.equipment) v += EQUIPMENT[k].cost * 0.6;
    return round2(v);
  }

  tierValue() {
    let v = 0;
    for (let i = 0; i < this.tier_idx; i++) v += TIERS[i].cost;
    return round2(v);
  }

  netWorth(market) {
    return round2(
      this.cash + this.inventoryValue(market) + this.equipmentValue() + this.tierValue() - this.debt
    );
  }

  // -- employees -----------------------------------------------------------

  hire(role, rng) {
    if (this.employees[role] || !EMPLOYEE_ROLE_KEYS.has(role)) return false;
    this.employees[role] = {
      skill: round1(rng.uniform(30, 60)),
      morale: 100.0,
      days: 0,
    };
    return true;
  }

  fire(role) {
    return delete this.employees[role];
  }

  dailyWages() {
    let w = 0;
    for (const role of Object.keys(this.employees)) w += EMPLOYEE_ROLES[role].salary;
    return round2(w);
  }

  staffBonus(kind) {
    // kind in service | quality | cleanliness | fuel. Pure — no mutation.
    const vals = {
      cashier: { service: 12, quality: 0, cleanliness: 0 },
      cook: { service: 0, quality: 6, cleanliness: 0 },
      cleaner: { service: 0, quality: 0, cleanliness: 20 },
      driver: { service: 0, quality: 0, cleanliness: 0 },
    };
    let total = 0.0;
    for (const [role, e] of Object.entries(this.employees)) {
      const eff = (e.skill / 50.0) * (0.7 + (0.3 * e.morale) / 100.0);
      total += vals[role][kind] * eff;
    }
    return total;
  }

  tickStaff() {
    for (const e of Object.values(this.employees)) {
      e.morale = Math.max(0.0, e.morale - 0.8);
      e.days += 1;
    }
  }

  snapshot() {
    return {
      cash: this.cash,
      loans: this.loans.map((l) => ({ principal: l.principal, rate: l.rate })),
      inventory: Object.fromEntries(Object.entries(this.inventory).map(([k, v]) => [k, { ...v }])),
      products: Object.fromEntries(Object.entries(this.products).map(([k, v]) => [k, { ...v }])),
      equipment: [...this.equipment].sort(),
      tech: [...this.tech].sort(),
      tier_idx: this.tier_idx,
      insurance: [...this.insurance].sort(),
      employees: Object.fromEntries(Object.entries(this.employees).map(([k, v]) => [k, { ...v }])),
      awareness: this.awareness,
      reputation: this.reputation,
      week_profit: this.week_profit,
      sales_tax_bill: this.sales_tax_bill,
      payroll_accum: this.payroll_accum,
      cleanliness: this.cleanliness,
      district: this.district,
      visited: [...this.visited].sort(),
    };
  }

  static restore(snap, difficulty) {
    const p = new Player(difficulty);
    p.cash = Number(snap.cash);
    p.loans = snap.loans.map((l) => new Loan(l.principal, l.rate));
    p.inventory = Object.fromEntries(Object.entries(snap.inventory).map(([k, v]) => [k, { qty: Number(v.qty), age: Number(v.age) }]));
    p.products = Object.fromEntries(Object.entries(snap.products).map(([k, v]) => [k, { qty: Number(v.qty), age: Number(v.age) }]));
    p.equipment = new Set(snap.equipment);
    p.tech = new Set(snap.tech);
    p.tier_idx = Number(snap.tier_idx);
    p.insurance = new Set(snap.insurance);
    p.employees = Object.fromEntries(Object.entries(snap.employees).map(([k, v]) => [k, { ...v }]));
    p.awareness = Number(snap.awareness);
    p.reputation = Number(snap.reputation);
    p.week_profit = Number(snap.week_profit);
    p.sales_tax_bill = Number(snap.sales_tax_bill);
    p.payroll_accum = Number(snap.payroll_accum);
    p.cleanliness = Number(snap.cleanliness);
    p.district = snap.district;
    p.visited = new Set(snap.visited);
    return p;
  }
}

import { EMPLOYEE_ROLES } from "./data.js";
const EMPLOYEE_ROLE_KEYS = new Set(Object.keys(EMPLOYEE_ROLES));
const round1 = (v) => Math.round(v * 10) / 10;
